Mathlib/Testing/SlimCheck/Sampleable.lean

```diff
@@ -0,0 +1,324 @@
+  f : α → β → Prop
+  total : ∀ x, ∃ y, f x y
+  determ : ∀ x y0 y1, f x y0 → f x y1 → y0 = y1
+
+namespace PhantomFunction
+
+/-- Evaluate the function hidden inside `ph`. -/
+noncomputable def eval (ph : PhantomFunction α β) (x : α) : β := Classical.choose $ ph.total x
+
+/-- Evaluating the function hidden inside a `PhantomFunction` is equivalent
+to evaluating the original function. -/
+@[simp]
+theorem eval_eq (ph : PhantomFunction α β) : ∀ (x : α) (y : β), ph.eval x = y ↔ ph.f x y := by
+  intro x y
+  simp [eval]
+  have h' := Classical.choose_spec <| ph.total x
+  constructor <;> intro h
+  next =>
+    rw [h] at h'
+    assumption
+  next =>
+    apply ph.determ x
+    exact h'
+    exact h
+
+/-- Hide a regular function inside a `PhantomFunction`. -/
+def mkPhantomFunction (g : α → β) : PhantomFunction α β where
+  f x y := g x = y
+  total x := ⟨_, rfl⟩
+  determ x y0 y1 h0 h1 := by rw [←h0, ←h1]
+
+/-- A specialized `mkPhantomFunction` for `SizeOf` instances. -/
+def mkSizeOf (α : Type u) [inst : SizeOf α] : PhantomFunction α Nat :=
+  mkPhantomFunction inst.sizeOf
+
+/-- A typeclass to contain `SizeOf` instances hidden inside a `PhantomFunction`
+unlike the auto generated `SizeOf` instances this one can be stored
+in datatypes for proof purposes. -/
+class PhantomSizeOf (α : Sort u) where
+  phSizeOf : PhantomFunction α Nat
+
+instance phantomSizeOf [SizeOf α] : PhantomSizeOf α where
+  phSizeOf := mkSizeOf α
+
+noncomputable instance sizeOfPhantom [ph : PhantomSizeOf α] : SizeOf α where
+  sizeOf := ph.phSizeOf.eval
+
+/-- A specialized `eval_eq` for `SizeOf` instances. -/
+@[simp]
+theorem sizeOfPhantom_eq_SizeOf [inst : SizeOf α] : @sizeOfPhantom α (@phantomSizeOf α inst) = inst := by
+  cases inst
+  unfold sizeOfPhantom
+  apply congrArg
+  funext x
+  rw [eval_eq]
+  rfl
+
+end PhantomFunction
+
+namespace SlimCheck
+
+open Random Gen PhantomFunction
+
+/-- `ShrinkFn α` is the type of functions that shrink an argument of type `α` -/
+abbrev ShrinkFn (α : Type u) [PhantomSizeOf α] := (x : α) → List { y : α // sizeOfPhantom.sizeOf y < sizeOfPhantom.sizeOf x }
+
+/-- `Sampleable α` provides ways of creating examples of type `α`. -/
+class Sampleable (α : Type u) [Repr α] where
+  sample {} : Gen α
+
+/-- Given an example `x : α`, `Shrinkable α` gives us a way to shrink it
+and suggest simpler examples. -/
+class Shrinkable (α : Type u) [phSz : PhantomSizeOf α] where
+
+  shrink : @ShrinkFn α phSz := λ _ => []
+/-- `SampleableExt` generalizes the behavior of `Sampleable` and `Shrinkable`
+and makes it possible to express instances for types that
+do not lend themselves to introspection, such as `ℕ → ℕ`.
+If we test a quantification over functions the
+counter-examples cannot be shrunken or printed meaningfully.
+For that purpose, `SampleableExt` provides a proxy representation
+`proxy` that can be printed and shrunken as well
+as interpreted (using `interp`) as an object of the right type. -/
+class SampleableExt (α : Sort u) where
```

Comment:
I realize that `SampleableExt` is already in the Lean 3 version so it's fine to just keep it as it is.

But I'm really not sure if this is a good API.  Any container type needs to have (duplicate?) Sampleable instances (this PR is missing `SampleableExt (α × β)`).  It would seem simpler to me to remove `Sampleable` and `Shrinkable` and only have `SampleableExt`.